### PR TITLE
Se 499 document amm fee buy

### PR DESCRIPTION
**Pull request by TJ-Everett**
Created document architecture for Fees, FEE-Guide.md, PROTOCOL-FEE-STRUCTURE.md, BUY-FEE.md, SELL-FEE.md

### CODE REVIEW

File: docs/userGuides/fees/AMM-BUY-FEE.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# AMM Buy Fee
+
+## Purpose
+
+The purpose of the AMM Buy Fee is to assess fees via the AMM swap function for purchases. A purchase via the AMM is defined as the user exchanging collateralized tokens for application tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM.
+
+Protocol supported tokens and AMMs will always assess all fees asigned to the account executing the current function. If a token and AMM have fees active and an account is tagged with applicable fees or a blank tag is used to assign a default fee, those fees are assessed on token transfers and AMM swaps (additive).
```
Comment (oscarsernarosero) [inline docs/userGuides/fees/AMM-BUY-FEE.md]: This is totally accurate with the code, but Im realizing there might be an issue with the way we implemented this. If we calculate fees on the `msg.sender` instead of the `from` or `to` address, then when a swap is executed, the `msg.sender` will be the AMM and not the owner of the tokens. Am I tripping?

File: docs/userGuides/fees/FEE-GUIDE.md
Comment (oscarsernarosero) [inline docs/userGuides/fees/FEE-GUIDE.md:1]: Maybe call this file README?

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+## Applies To:
+
+- [x] ERC20
+- [x] ERC721
+- [x] AMM
+
+## Scope 
+Token Fees work at the token level. Fees must be activated and configured for each token in the corresponding token handler. Token fees are assessed in the transfer function of the token. 
+
+AMM Fees work at the AMM level. Fees must be activated and configured for each AMM in the corresponding AMM handler. AMM fees are assessed in the swap function of the AMM.
```
Comment (oscarsernarosero) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Maybe create a subsection for each kind, or use bullets.

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+Token Evaluation: 
+The token determines if the handler has fees active. 
+The token retrieves all applicable fees for the account transferring tokens (msg.sender).
+The token loops through each applicable fee and sends that amount from the transfer total to the `feeCollectorAccount` for that fee. The total amount of fees assesed is tracked within the transfer as `fees`, upon complettion of the loop the amount of tokens minus the `fees` is transferred to the recipient of the transaction.  
+
+AMM Evaluation: 
+The AMM determines if the AMM handler has fees active.
+The AMM retrieves all applicable fees for the account transferring tokens (swpper) for the token being swapped. If token 1 is swapped for token 0, token 1 has fees assessed. 
+The AMM loops through each applicable fee and sends that amount from the transfer total to the `feeCollectorAccount` for that fee. The total amount of fees assesed is tracked within the swap as `fees`, upon complettion of the loop the amount of tokens in minus the `fees` is transferred to the AMM and the amount of tokens out are transferred to the user.
```
Comment (oscarsernarosero) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: same here. Probably subsection is better.

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+###### *see [ProtocolERC20](../../../src/client/token/ERC20/ProtocolERC20.sol) -> transfer*
+
+## Evaluation Exceptions 
+- There are no evaluation exceptions when fees are active.
```
Comment (oscarsernarosero) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: what about treasury?

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+## Return Data
+
+When assessing fees the function getApplicableFees() returns: 
+- **feeCollectorAccounts** (address[]): List of fee recipient addresses
+- **feePercentagess** (int24[]): List of fee percentages 
+
+## Data Recorded
+
+Fee totals are added inside of a loop and then total fees are subtracted from the amount being transferred within the transfer or swap functions. This data is not saved to storage.
```
Comment (oscarsernarosero) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md:167]: nice

**Review by oscarsernarosero — CHANGES_REQUESTED** [review]
This is a great job Tayler. I like the structure. I requested some minor changes, but I have always been a bit confused about fees, so I might be off on some comments.

File: docs/userGuides/fees/AMM-BUY-FEE.md (hunk 1 above)
Comment (mpetersoCode55) [inline docs/userGuides/fees/AMM-BUY-FEE.md]: Discussed in huddle, it may be worth adding a note that token transfer fees are taken from the protocol supported token being transferred (instead of a collateralized token like in the AMM swap)

File: docs/userGuides/fees/AMM-SELL-FEE.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# AMM Sell Fee
+
+## Purpose
+
+The purpose of the AMM Sell Fee is to assess fees via the AMM swap function for purchases. A purchase via the AMM is defined as the user exchanging application tokens for collateralized tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM.
+
+Protocol supported tokens and AMMs will always assess all fees asigned to the account executing the current function. If a token and AMM have fees active and an account is tagged with applicable fees or a blank tag is used to assign a default fee, those fees are assessed on token transfers and AMM swaps (additive).
```
Comment (mpetersoCode55) [inline docs/userGuides/fees/AMM-SELL-FEE.md]: Same note from the Buy Fee

File: docs/userGuides/fees/FEE-GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+| [AMM Buy Fee](./AMM-BUY-FEE.md)  | The purpose of the AMM Buy Fee is to assess fees via the AMM swap function for purchases. A purchase via the AMM is defined as the user exchanging collateralized tokens for application tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM. | 
+| [AMM Sell Fee](./AMM-SELL-FEE.md) | The purpose of the AMM Sell Fee is to assess fees via the AMM swap function for purchases. A purchase via the AMM is defined as the user exchanging application tokens for collateralized tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM. |
```
Comment (mpetersoCode55) [inline docs/userGuides/fees/FEE-GUIDE.md]: Shouldn't this one be focused on the sale side instead of the purchase? Just reading the two overviews I can't tell the difference between AMM Buy Fee and AMM Sell Fee

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+///     tag   => Fee struct 
+mapping(bytes32 => Fee) feesByTag;
+```
+
+###### *see [Fees](../../../src/client/token/data/Fees.sol)*
+
+## Configuration and Enabling/Disabling
+- Fees can only be configured in the asset handler by the **owner**.
+- Fees can only be added in the asset handler by the **owner**.
+- Fees can only be aremoved in the asset handler by the **owner**.
```
Comment (mpetersoCode55) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: typo on removed

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+## Fees Evaluation
+
+Token Evaluation: 
+The token determines if the handler has fees active. 
+The token retrieves all applicable fees for the account transferring tokens (msg.sender).
+The token loops through each applicable fee and sends that amount from the transfer total to the `feeCollectorAccount` for that fee. The total amount of fees assesed is tracked within the transfer as `fees`, upon complettion of the loop the amount of tokens minus the `fees` is transferred to the recipient of the transaction.  
+
+AMM Evaluation: 
+The AMM determines if the AMM handler has fees active.
+The AMM retrieves all applicable fees for the account transferring tokens (swpper) for the token being swapped. If token 1 is swapped for token 0, token 1 has fees assessed.
```
Comment (mpetersoCode55) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Is this true? I thought the AMM fees were always take from the collateralized token, regardless of which direction you're swapping.

**Review by mpetersoCode55 — CHANGES_REQUESTED** [review]

File: docs/userGuides/fees/FEE-GUIDE.md
Comment (TJ-Everett) [reply docs/userGuides/fees/FEE-GUIDE.md:1]: FEE-README.md?

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md (hunk 4 above)
Comment (TJ-Everett) [reply docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Great Question! Since fees will be assessed on every transfer or swap when the fee rules are active- there isn't an "exception" to the check but those would accounts and contracts wouldn't be tagged with a fee so they don't have a fee but are still checked. 

Should I explain that in this section?

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md (hunk 9 above)
Comment (TJ-Everett) [reply docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Great catch I will reword this to a true statement.

File: docs/userGuides/fees/AMM-BUY-FEE.md (hunk 1 above)
Comment (TJ-Everett) [reply docs/userGuides/fees/AMM-BUY-FEE.md]: You mean that in a scenario where there are token fees and AMM fees we would not be assessing the token fee in the swap because we pass in msg.sender instead of from/to? 

Inside the AMM the msg.sender is the address calling swap. Then that address is used throughout. I don't think it is is an issue, outside of the possibility above. Unless I am misunderstanding this entirely.

File: docs/userGuides/fees/FEE-GUIDE.md
Comment (mpetersoCode55) [reply docs/userGuides/fees/FEE-GUIDE.md:1]: Or just shorten it to FEEDME.md

**Review by mpetersoCode55 — APPROVED** [review]

File: docs/userGuides/fees/AMM-BUY-FEE.md (hunk 1 above)
Comment (oscarsernarosero) [reply docs/userGuides/fees/AMM-BUY-FEE.md]: I think you're right.

File: docs/userGuides/fees/FEE-GUIDE.md
Comment (oscarsernarosero) [reply docs/userGuides/fees/FEE-GUIDE.md:1]: Well, the reason to call it README.md specifically is so it can automatically be shown by GH page once you land on this directory. Since this is basically a map of the directory, it makes sense for this document to be the one showing up.

**Review by oscarsernarosero — APPROVED** [review]
Well done! I'll leave it to you if you want to make the FEE-GUIDE the page popping up.

File: docs/userGuides/fees/AMM-BUY-FEE.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# AMM Buy Fee
+
+## Purpose
+
+The purpose of the AMM Buy Fee is to assess fees via the AMM swap function for purchases. A purchase via the AMM is defined as the user exchanging collateralized tokens for application tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-BUY-FEE.md]: I think the MATIC is always just referred to as MATIC. I've never seen it abbreviated MATC.

File: docs/userGuides/fees/AMM-BUY-FEE.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+Protocol supported tokens and AMMs will always assess all fees asigned to the account executing the current function. If a token and AMM have fees active and an account is tagged with applicable fees or a blank tag is used to assign a default fee, those fees are assessed on token transfers and AMM swaps (additive). Token fees are assessed and taken from the token itself, not a collateralized token, when fees are active in the token handler. 
+
+
+## Applies To:
+
+- [x] ERC20
+- [x] ERC721
+- [x] AMM
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-BUY-FEE.md]: I'm not sure this really applies to all 3 here. This document is titled, "AMM-BUY-FEE". This "Applies To:" section seems a little confusing to me. Can we get rid of it?

File: docs/userGuides/fees/AMM-BUY-FEE.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+## Application of Buy Fee
+ 
+A Buy Fee is the additive amount of all applicable fees for the account calling the swap function that swaps the user's collateralized tokens for application tokens inside the AMM.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-BUY-FEE.md]: This is a complex sentence. Can we break it up into pieces like, "Buy fees are assessed on the collateralized token being swapped for application tokens inside the AMM. The fees are additive. If a user has multiple applicable fees, the sum of all their fees will be assessed."

File: docs/userGuides/fees/AMM-SELL-FEE.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# AMM Sell Fee
+
+## Purpose
+
+The purpose of the AMM Sell Fee is to assess fees via the AMM swap function for sales. A sale via the AMM is defined as the user exchanging application tokens for collateralized tokens via the AMM. Collateralized tokens can be chain native tokens (ETH/MATC), non protocol supported ERC20 tokens (WETH/USDC) or protocol supported ERC20 tokens. AMM Fees are always taken from the collateralized token within the AMM.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-SELL-FEE.md]: Same comment about MATC as above.

File: docs/userGuides/fees/AMM-SELL-FEE.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+Protocol supported tokens and AMMs will always assess all fees asigned to the account executing the current function. If a token and AMM have fees active and an account is tagged with applicable fees or a blank tag is used to assign a default fee, those fees are assessed on token transfers and AMM swaps (additive). Token fees are assessed and taken from the token itself, not a collateralized token, when fees are active in the token handler. 
+
+
+
+## Applies To:
+
+- [x] ERC20
+- [x] ERC721
+- [x] AMM
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-SELL-FEE.md]: Same comment as above.

File: docs/userGuides/fees/AMM-SELL-FEE.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+## Application of Sell Fee
+ 
+A Sell Fee is the additive amount of all applicable fees for the account calling the swap function that swaps the user's application tokens for collateralized tokens inside the AMM.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/AMM-SELL-FEE.md]: This is a complex sentence. Can we break it up into pieces like, "Sell fees are assessed on the collateralized token being received during swaps for application tokens inside the AMM. The fees are additive. If a user has multiple applicable fees, the sum of all their fees will be assessed."

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+///     tag   => Fee struct 
+mapping(bytes32 => Fee) feesByTag;
+```
+
+###### *see [Fees](../../../src/client/token/data/Fees.sol)*
+
+## Configuration and Enabling/Disabling
+- Fees can only be configured in the asset handler by the **owner**.
+- Fees can only be added in the asset handler by the **owner**.
+- Fees can only be removed in the asset handler by the **owner**.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Fees are configured by the Rule Admins. The Fee contract has owner on it but the owner is actually the handler.

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+## Fees Evaluation
+
+### Token Evaluation: 
+The token determines if the handler has fees active. 
+The token retrieves all applicable fees for the account transferring tokens (msg.sender).
+The token loops through each applicable fee and sends that amount from the transfer total to the `feeCollectorAccount` for that fee. The total amount of fees assesed is tracked within the transfer as `fees`, upon complettion of the loop the amount of tokens minus the `fees` is transferred to the recipient of the transaction.  
+
+### AMM Evaluation: 
+The AMM determines if the AMM handler has fees active.
+The AMM retrieves all applicable fees for the account calling the swap function (swpper).
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: swpper?

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+
+The selector for this error is `0x248ee764`.
+
+
+## Add Function
+
+Adding a fee is done through the function:
+
+```c
+function addFee(bytes32 _tag, uint256 _minBalance, uint256 _maxBalance, int24 _feePercentage, address _targetAccount) external onlyOwner;
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: This function should be the handler's pass through function. The function in Fees.sol will never be called directly.

File: docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+### Parameters:
+
+- **_tag** (bytes32): tag for fee application to an account.
+- **minBalance** (uint256): minimum balance for fee application 
+- **maxBalance** (uint256): maximum balance for fee application 
+- **feePercentage** (int24): fee percentage to assess in basis units (-10000 to 10000)
+- **_targetAccount** (address): address of the fees recipient account 
+
+The create function will return the protocol ID of the rule.
```
Comment (ShaneDuncan602) [inline docs/userGuides/fees/PROTOCOL-FEE-STRUCTURE.md]: Fees don't have protocol ID's. The FeeRuleDataFacet still exists but is not currently used.

**Review by ShaneDuncan602 — CHANGES_REQUESTED** [review]
Just a few minor things. This is good stuff. A lot of good detail.

**Review by ShaneDuncan602 — APPROVED** [review]
Looks perfect!
